tests/test_pure_pursuit.py

```diff
@@ -1,17 +1,84 @@
-import utilities.pure_pursuit as pp
+from utilities.pure_pursuit import PurePursuit
+import utilities.pure_pursuit as p_p
+from hypothesis.strategies import floats
+from hypothesis import given
+import math
 
 
 def test_trapezoidal():
-    waypoints = [pp.Waypoint(0, 0, 0, 0), pp.Waypoint(10, 10, 0, 2)]
-    trap = pp.insert_trapezoidal_waypoints(waypoints, 2, -2)
+    waypoints = [p_p.Waypoint(0, 0, 0, 0), p_p.Waypoint(10, 10, 0, 2)]
+    trap = p_p.insert_trapezoidal_waypoints(waypoints, 2, -2)
     assert len(trap) == 3
     assert trap[1][2] == waypoints[1][2], (
         "Intermediate waypoint should have end speed when accelerating: %s" % trap
     )
 
-    waypoints = [pp.Waypoint(10, 10, 0, 2), pp.Waypoint(0, 0, 0, 0)]
-    trap = pp.insert_trapezoidal_waypoints(waypoints, 2, -2)
+    waypoints = [p_p.Waypoint(10, 10, 0, 2), p_p.Waypoint(0, 0, 0, 0)]
+    trap = p_p.insert_trapezoidal_waypoints(waypoints, 2, -2)
     assert len(trap) == 3
     assert trap[1][2] == waypoints[0][2], (
         "Intermediate waypoint should have beginning speed when decelerating: %s" % trap
     )
+
+
+@given(
+    floats(min_value=0, max_value=2),
+    floats(min_value=0, max_value=5),
+    floats(min_value=0, max_value=5),
+)
+def test_find_speed(distance_along_path, start_speed, end_speed):
+    pp = PurePursuit(look_ahead=0.2, look_ahead_speed_modifier=0.25)
+    start_path_distance = 0
+    end_path_distance = 2
+
+    speed = pp.find_speed(
+        start_path_distance,
+        end_path_distance,
+        start_speed,
+        end_speed,
+        distance_along_path,
+    )
+    if end_speed > start_speed:
+        assert speed <= end_speed
+        assert speed >= start_speed
+    if end_speed < start_speed:
+        assert speed >= end_speed - 1e-8
+        assert speed <= start_speed + 1e-8
+
+
+@given(floats(min_value=0, max_value=10))
+def test_distance_along_path(x):
+    pp = PurePursuit(look_ahead=0.2, look_ahead_speed_modifier=0.25)
+    pp.last_robot_x = 0
+    pp.last_robot_y = 0
+    robot_position = (x, 0)
+    distance = pp.distance_along_path(robot_position)
+
+    assert math.isclose(distance, x, rel_tol=1e-8)
+
+
+@given(
+    floats(min_value=0, max_value=9),
+    floats(min_value=-4, max_value=4),
+    floats(min_value=-4, max_value=4),
+)
+def test_find_velocity(pos_x, start_spd, end_spd):
+    pp = PurePursuit(look_ahead=0.2, look_ahead_speed_modifier=0.25)
+    pp.last_robot_x = 0
+    pp.last_robot_y = 0
+    robot_position = (pos_x, 0)
+    pp.waypoints = [
+        p_p.Segment(0, 0, 0, start_spd, 0),
+        p_p.Segment(10, 0, 0, end_spd, 10),
+    ]
+    vx, vy, _ = pp.find_velocity(robot_position)
+    assert 1 <= 2 <= 3
```

Comment:
This should definitely be always true... but I'm not sure this is relevant in our tests.